### ISSUE TITLE
Non-standard european Swaption no longer pricing with LGM in master

### ISSUE THREAD

**Issue by noonediesalone** [OpenSourceRisk/Engine#223]
Hi,

Comparing master to latest release 11, I see non-standard european Swaptions (e.g. non-constant notional/rate/spread/gearing) no longer use LGM.

Looking here: https://github.com/OpenSourceRisk/Engine/blob/v1.8.11.0/OREData/ored/portfolio/swaption.cpp
```
    if (exerciseType_ == Exercise::European && isStandard)
        buildEuropean(engineFactory);
    else
        buildBermudan(engineFactory);
```
and here: https://github.com/OpenSourceRisk/Engine/blob/master/OREData/ored/portfolio/swaption.cpp
```
    std::string builderType;
    std::vector<std::string> builderPrecheckMessages;
    // for European swaptions that are not handled by the black pricer, we fall back to the numeric Bermudan pricer
    if (exerciseType_ == Exercise::European &&
        QuantExt::BlackMultiLegOptionEngineBase::instrumentIsHandled(*swaption, builderPrecheckMessages)) {
        builderType = "EuropeanSwaption";
    } else {
        QL_REQUIRE(
            QuantExt::NumericLgmMultiLegOptionEngineBase::instrumentIsHandled(*swaption, builderPrecheckMessages),
            "Swaption::build(): instrument is not handled by the available engines: " +
                boost::join(builderPrecheckMessages, ", "));
        if (exerciseType_ == Exercise::Bermudan)
            builderType = "BermudanSwaption";
        else if (exerciseType_ == Exercise::American)
            builderType = "AmericanSwaption";
    }

    auto swaptionBuilder = boost::dynamic_pointer_cast<SwaptionEngineBuilder>(engineFactory->builder(builderType));
    QL_REQUIRE(swaptionBuilder, "Swaption::build(): internal error: could not cast to SwaptionEngineBuilder");
```
.. and `QuantExt::BlackMultiLegOptionEngineBase::instrumentIsHandled()` doesn't look at non-constant notional/rate/spread/gearing .. which suggests that `calculate()` should handle all these cases.

But from our tests, the valuation is extremely off for a notional accretion callable swap with a single call date.

An easy workaround would be to just make the Swaption as Bermudan and will follow similar path as in release 11.

Just asking here if this is intended behavior, from the model choice, or more like a bug?

Regards,
Laurentiu.

**Comment by pcaspers** [comment]
I'll take a look. Would you mind posting a trade xml producing a large difference?

**Comment by noonediesalone** [comment]
Hi Peter,

Thank you for picking up this case 😄 

Please use this sample Swap + Swaption: [portfolio.zip](https://github.com/OpenSourceRisk/Engine/files/14478918/portfolio.zip)

You can just drop it in `Example_5\Input` and run npv with build from release version 11 vs. master.
It's supposed to be a "zero coupon" swap with accretion notional, e.g. original notional is 30 mio with 1 mio increase each year which is offset by the fixed coupon (about 3.33% annual rate) and the option which has Physical settlement has expiry date relatively close to present but with 10 more year in the underlying.

Swap gives 8,767,203 while the Swaption gives -8,817,857 in release vs. -49,470,657 in master.
If we switch the Style to Bermudan, both will use LGM and agree on the valuation from release version, which is correct and nets the swap.

Regards,
Laurentiu.

**Comment by pcaspers** [comment]
Hi Laurentiu,

thanks for highlighting this issue. I notice there are actually a couple of issues with the newly introduced BlackMultiLegOptionEngine, which we will fix soon - at the latest with the release 12 which is planned to come out within the next weeks. However I think the issue you see is mainly due to this line

https://github.com/OpenSourceRisk/Engine/blob/master/QuantExt/qle/pricingengines/blackmultilegoptionengine.cpp#L254

which should be

```
            simpleCfCorrection += c.amount * c.discountFactor / fixedBps;
```

i.e. there is a missing "+" sign. With that fix I get an NPV of -8,817,252.474243 for the swaption, which is quite close to what we previously computed in the LGM setting.

I would generally like to keep the new engine assignment, i.e. if a swaption is entered as "European" then the new BlackMultiLegOptionEngine is used except there are features that can not be handled (like cap / floors in the underlying). Note the following points:

- as you already mentioned, you can always set up the swaption as "Bermudan" without changing the trade economics. This will enforce the usage of the LGM engine configured under product type BermudanSwaption in the pricing engine config
- you can now also use the LGM model for product type EuropeanSwaption in the pricing engine configuration, since the underlying instruments are compatible now (all swaptions are represented as MultiLegOption internally)

What we can think about is to refine the product assignment, i.e. instead of just having EuropeanSwaption and BermudanSwaption in pricingengine xml we could differentiate more subtypes of swaptions like acreeting, amortizing swaptions, etc. with separate model / engine configs for each of them, if that is useful.

Best
Peter

**Comment by noonediesalone** [comment]
Thanks Peter for your insights!

I like the flexibility that your proposal brings to the model choice.
Maybe, as one suggestion, to have an optional field in OptionData that overrides the model choice (e.g. BermudanSwaption, EuropeanSwaption, AmortizingSwaption, etc.). When that is provided (and there is a pricing definition for this "product" type) this choice will override the default introspection logic (which looks inside the trade and figures what's the right model to use).
I think generally for IR Swaptions are enough 4 pricing definitions: one for Bermudan, one for American and two for European (one standard and one generalized).

Will follow your developments in the new release.

Regards,
Laurentiu.

**Comment by noonediesalone** [comment]
Hi Peter,

Yes, envelope might be a better place, since everybody can access it. 
Maybe downside is with multi-leg swaps (e.g. floating + fixed + CMS), where each leg calls for it's own product pricer. In this case maybe it doesn't make sense because each leg will be one product which maps into one pricer and there is no choice.

Regards,
Laurentiu.

**Comment by pcaspers** [comment]
Good point! I think in this case I'd use `pricing_product_type_1`, `pricing_product_type_2` ... in addition to override coupon pricer for leg #1, #2, ....
